src/commands/git-metadata/gitdb.ts

```diff
@@ -0,0 +1,281 @@
+  } catch (err) {
+    log.warn(`Failed getting commits to upload: ${err}`)
+    throw err
+  }
+
+  let packfiles
+  try {
+    packfiles = generatePackFilesForCommits(log, commitsToUpload)
+    log.debug(`${packfiles.length} packfiles generated.`)
+  } catch (err) {
+    log.warn(`Failed generating packfiles: ${err}`)
+    throw err
+  }
+
+  if (dryRun) {
+    log.debug(`Dry-run enabled, not uploading anything.`)
+
+    return
+  }
+  log.debug(`Uploading packfiles...`)
+  try {
+    await uploadPackfiles(log, request, repoURL, latestCommits[0], packfiles)
+    log.debug(`Successfully uploaded packfiles.`)
+  } catch (err) {
+    log.warn(`Failed to upload packfiles: ${err}`)
+    throw err
+  }
+}
+
+const getRepoURL = gitRemote
+
+const getLatestLocalCommits = async (git: simpleGit.SimpleGit) => {
+  const logResult = await git.log(['-n 1000', '--since="1 month ago"'])
+
+  return logResult.all.map((c) => c.hash)
+}
+
+// getKnownCommits asks the backend which of the given commits are already known
+const getKnownCommits = async (log: Logger, request: RequestBuilder, repoURL: string, latestCommits: string[]) => {
+  interface SearchCommitResponse {
+    data: Commit[]
+  }
+
+  interface Commit {
+    type: string
+    id: string
+  }
+
+  const localCommitData = JSON.stringify({
+    meta: {
+      repository_url: repoURL,
+    },
+    data: latestCommits.map((commit) => ({
+      id: commit,
+      type: 'commit',
+    })),
+  })
+  const response = await runRequest(log, 'search_commits', () =>
+    request({
+      url: '/api/v2/git/repository/search_commits',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      data: localCommitData,
+      method: 'POST',
+      timeout: API_TIMEOUT,
+    })
+  )
+  const commits = response.data as SearchCommitResponse
+  if (!commits || commits.data === undefined) {
+    throw new Error(`Invalid API response: ${response}`)
+  }
+
+  return commits.data.map((c) => {
+    if (c.type !== 'commit' || c.id === undefined) {
+      throw new Error('Invalid commit type response')
+    }
+
+    return sanitizeCommit(c.id)
+  })
+}
+
+const sanitizeCommit = (sha: string) => {
+  const isValidSha = (s: string) => /[0-9a-f]{40}/.test(s)
+
+  const sanitizedCommit = sha.replace(/[^0-9a-f]+/g, '')
+  if (sanitizedCommit !== sha) {
+    throw new Error(`Invalid commit format: ${sha} (different from sanitized ${sanitizedCommit})`)
+  }
+  if (!isValidSha(sanitizedCommit)) {
+    throw new Error(`Invalid commit format: ${sanitizedCommit}`)
+  }
+
+  return sanitizedCommit
+}
+
+const getCommitsToUpload = async (git: simpleGit.SimpleGit, commitsToExclude: string[]) => {
+  const rawResponse = await git.raw(
+    ['rev-list', '--objects', '--no-object-names', '--filter=blob:none', '--since="1 month ago"', 'HEAD'].concat(
+      commitsToExclude.map((sha) => '^' + sha)
+    )
+  )
+  const commitsToInclude = rawResponse.split('\n').filter((c) => c !== '')
+
+  return commitsToInclude
```

Comment:
You are bringing into memory all of the ids of all objects that need to be uploaded. That can potentially be megabytes worth of data. I'm not sure if the tracers are bringing this info to memory or if they are pipeline the `git rev-list` command to the `git pack-objects` command.

I am fine with this as a first approach, but I thought it would be good to mention that this can be memory intensive.

Reply:
My understanding is that at least the dd-trace-js implementation is loading it in memory as well, see [here](https://github.com/DataDog/dd-trace-js/blob/6763246af16adc2e00b292410241df6be14817db/packages/dd-trace/src/plugins/util/git.js#L50).

AFAIK this should take 40 bytes * (number of commits to upload + number of tree objects to upload). I expect it to be a few MBs max, since we're only sending data from the past month, but I might be wrong? If my expectations are right I'd expect it to be fine for most use cases?

Anyway it's indeed worth adding a comment.

Reply:
Yeah, I think it is fine leaving as is.